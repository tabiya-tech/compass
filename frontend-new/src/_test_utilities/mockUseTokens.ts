export function mockUseTokens(overrides = {}) {
  jest.spyOn(require("src/auth/hooks/useTokens"), "useTokens").mockImplementation(() => {
    return {
      isAuthenticating: false,
      isAuthenticated: true,
      setIsAuthenticated: jest.fn(),
      setToken: jest.fn(),
      clearTokens: jest.fn(),
      ...overrides,
    };
  });
}
