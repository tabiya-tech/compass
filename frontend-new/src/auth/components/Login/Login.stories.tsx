import React from "react";
import { Meta, StoryObj } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import Login from "./Login";
import { AuthContext, authContextDefaultValue } from "src/auth/AuthProvider/AuthProvider";

const meta: Meta<typeof Login> = {
  title: "Auth/Login",
  tags: ["autodocs"],
  component: Login,
  decorators: [
    (Story) => {
      const mockAuthContextValue = {
        ...authContextDefaultValue,
        login: action("login"),
      };

      return (
        <AuthContext.Provider value={mockAuthContextValue}>
          <Story />
        </AuthContext.Provider>
      );
    },
  ],
};

export default meta;

export const Shown: StoryObj<typeof Login> = {
  args: {},
};

export const LoggingIn: StoryObj<typeof Login> = {
  args: {},
  render: (props) => {
    return (
      <AuthContext.Provider
        value={{
          ...authContextDefaultValue,
          isLoggingIn: true,
        }}
      >
        <Login {...props} />
      </AuthContext.Provider>
    );
  },
};
