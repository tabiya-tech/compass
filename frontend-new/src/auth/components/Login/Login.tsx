import React, { useCallback, useContext, useState } from "react";
import { Container, Box, TextField, Button, Typography, useTheme, styled } from "@mui/material";
import { AuthContext, TabiyaUser } from "src/auth/AuthProvider";
import { NavLink as RouterNavLink, useNavigate } from "react-router-dom";
import { routerPaths } from "src/app/routerPaths";
import IDPAuth from "src/auth/components/IDPAuth/IDPAuth";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import UserPreferencesService from "src/auth/services/UserPreferences/userPreferences.service";
import AuthContextMenu from "src/auth/components/AuthContextMenu/AuthContextMenu";

const uniqueId = "7ce9ba1f-bde0-48e2-88df-e4f697945cc4";

const StyledNavLink = styled(RouterNavLink)(({ theme }) => ({
  color: theme.palette.text.textAccent,
  fontStyle: "italic",
  textDecoration: "none",
  "&:hover": {
    textDecoration: "underline",
  },
}));

export const DATA_TEST_ID = {
  LOGIN_CONTAINER: `login-container-${uniqueId}`,
  LOGO: `login-logo-${uniqueId}`,
  TITLE: `login-title-${uniqueId}`,
  SUBTITLE: `login-subtitle-${uniqueId}`,
  FORM: `login-form-${uniqueId}`,
  EMAIL_INPUT: `login-email-input-${uniqueId}`,
  PASSWORD_INPUT: `login-password-input-${uniqueId}`,
  LOGIN_BUTTON: `login-button-${uniqueId}`,
  FORGOT_PASSWORD_LINK: `login-forgot-password-link-${uniqueId}`,
  LOGIN_USING: `login-using-${uniqueId}`,
  FIREBASE_AUTH_CONTAINER: `firebase-auth-container-${uniqueId}`,
  LOGIN_LINK: `login-login-link-${uniqueId}`,
  LANGUAGE_SELECTOR: `login-language-selector-${uniqueId}`,
};

const Login: React.FC = () => {
  const theme = useTheme();
  const { login } = useContext(AuthContext);
  const navigate = useNavigate();
  const { enqueueSnackbar } = useSnackbar();

  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");

  /**
   * Check if the user has accepted the terms and conditions
   * @param user
   */
  const checkUserPreferences = useCallback(
    async (user: TabiyaUser) => {
      const userPreferencesService = new UserPreferencesService();
      try {
        const userPreferences = await userPreferencesService.getUserPreferences(user.id);
        const acceptedTcDate = new Date(userPreferences.accepted_tc);
        // If the accepted_tc is not set or is not a valid date, redirect to the DPA page
        // this is to ensure that even if the accepted_tc is manipulated in the database, the user will be redirected to the DPA page
        // and will have to accept the terms and conditions again
        if (!userPreferences.accepted_tc || isNaN(acceptedTcDate.getTime())) {
          navigate(routerPaths.DPA, { replace: true });
        } else {
          navigate(routerPaths.ROOT, { replace: true });
        }
      } catch (e) {
        enqueueSnackbar("Failed to fetch user preferences", { variant: "error" });
        console.error("Failed to fetch user preferences", e);
      }
    },
    [navigate, enqueueSnackbar]
  );

  /**
   * Handle the login form submission
   * @param event
   */
  const handleLogin = async (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    login(
      email,
      password,
      async (user) => {
        await checkUserPreferences(user);
        enqueueSnackbar("Login successful", { variant: "success" });
      },
      (error) => {
        if (error.message === "Email not verified") {
          enqueueSnackbar("Please verify your email", { variant: "error" });
          return;
        }
        console.error("Login failed", error);
        enqueueSnackbar("Login failed", { variant: "error" });
      }
    );
  };

  return (
    <Container maxWidth="xs" sx={{ height: "100%" }} data-testid={DATA_TEST_ID.LOGIN_CONTAINER}>
      <Box
        display="flex"
        flexDirection="column"
        alignItems="center"
        justifyContent={"space-evenly"}
        m={4}
        height={"80%"}
      >
        <Box display="flex" flexDirection="row" justifyContent="center" alignItems="center">
          <img
            src={`${process.env.PUBLIC_URL}/logo.svg`}
            alt="Logo"
            style={{ maxWidth: "60%", margin: "10%" }}
            data-testid={DATA_TEST_ID.LOGO}
          />
          <AuthContextMenu />
        </Box>
        <Typography variant="h4" gutterBottom data-testid={DATA_TEST_ID.TITLE}>
          Welcome to Compass!
        </Typography>
        <Typography variant="body2" gutterBottom data-testid={DATA_TEST_ID.SUBTITLE}>
          Login to your account to continue
        </Typography>
        <Box component="form" mt={2} onSubmit={handleLogin} data-testid={DATA_TEST_ID.FORM}>
          <TextField
            fullWidth
            label="Email"
            type="email"
            variant="outlined"
            margin="normal"
            required
            onChange={(e) => setEmail(e.target.value)}
            inputProps={{ "data-testid": DATA_TEST_ID.EMAIL_INPUT }}
          />
          <TextField
            fullWidth
            label="Password"
            type="password"
            variant="outlined"
            margin="normal"
            required
            onChange={(e) => setPassword(e.target.value)}
            inputProps={{ "data-testid": DATA_TEST_ID.PASSWORD_INPUT }}
          />
          <Button
            fullWidth
            variant="contained"
            color="primary"
            style={{ marginTop: 16 }}
            type="submit"
            data-testid={DATA_TEST_ID.LOGIN_BUTTON}
          >
            Login
          </Button>
        </Box>
        <Box
          display="flex"
          flexDirection="column"
          justifyContent="center"
          alignItems="center"
          mt={(theme) => theme.tabiyaSpacing.lg}
          data-testid={DATA_TEST_ID.FIREBASE_AUTH_CONTAINER}
        >
          <Typography variant="body2" mt={2} data-testid={DATA_TEST_ID.LOGIN_USING}>
            Or continue with
          </Typography>
          <Box mt={2} width="100%">
            <IDPAuth />
          </Box>
        </Box>
        <Typography variant="body2" mt={2} data-testid={DATA_TEST_ID.LOGIN_LINK}>
          Dont have an account?{" "}
          <StyledNavLink
            to={routerPaths.REGISTER}
            style={{
              color: theme.palette.text.textAccent,
              fontStyle: "italic",
            }}
          >
            Register
          </StyledNavLink>
        </Typography>
      </Box>
    </Container>
  );
};

export default Login;
