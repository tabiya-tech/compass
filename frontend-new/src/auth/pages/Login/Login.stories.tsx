import { Meta, StoryObj } from "@storybook/react";
import Login from "./Login";

const meta: Meta<typeof Login> = {
  title: "Auth/Login",
  tags: ["autodocs"],
  component: Login,
};

export default meta;

export const Shown: StoryObj<typeof Login> = {
  args: {},
};
