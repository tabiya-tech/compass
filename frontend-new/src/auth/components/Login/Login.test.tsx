import "src/_test_utilities/consoleMock";
import React from "react";
import { render, screen, waitFor, fireEvent } from "src/_test_utilities/test-utils";
import { HashRouter, useNavigate } from "react-router-dom";
import Login, { DATA_TEST_ID } from "./Login";
import { AuthContext, TabiyaUser } from "src/auth/Providers/AuthProvider/AuthProvider";
import { useSnackbar } from "src/theme/SnackbarProvider/SnackbarProvider";
import { routerPaths } from "src/app/routerPaths";
import { mockUseTokens } from "src/_test_utilities/mockUseTokens";
import { ServiceError } from "src/error/error";
import ErrorConstants from "src/error/error.constants";
import ErrorCodes = ErrorConstants.ErrorCodes;
import { StatusCodes } from "http-status-codes";
import { UserPreferencesContext } from "src/auth/Providers/UserPreferencesProvider/UserPreferencesProvider";
import { Language } from "src/auth/services/UserPreferences/userPreferences.types";
import * as Logger from "src/error/logger";

// Mock the envService module
jest.mock("src/envService", () => ({
  getFirebaseAPIKey: jest.fn(() => "mock-api-key"),
  getFirebaseDomain: jest.fn(() => "mock-auth-domain"),
  getBackendUrl: jest.fn(() => "mock-backend-url"),
}));

//mock the IDPAuth component
jest.mock("src/auth/components/IDPAuth/IDPAuth", () => {
  const actual = jest.requireActual("src/auth/components/IDPAuth/IDPAuth");
  return {
    ...actual,
    __esModule: true,
    default: jest.fn().mockImplementation(() => {
      return <span data-testid={actual.DATA_TEST_ID.FIREBASE_AUTH_CONTAINER}></span>;
    }),
  };
});

// mock the snack bar provider
jest.mock("src/theme/SnackbarProvider/SnackbarProvider", () => {
  const actual = jest.requireActual("src/theme/SnackbarProvider/SnackbarProvider");
  return {
    ...actual,
    __esModule: true,
    useSnackbar: jest.fn().mockReturnValue({
      enqueueSnackbar: jest.fn(),
      closeSnackbar: jest.fn(),
    }),
  };
});

// mock the router
jest.mock("react-router-dom", () => {
  const actual = jest.requireActual("react-router-dom");
  return {
    ...actual,
    __esModule: true,
    useNavigate: jest.fn().mockReturnValue(jest.fn()),
    NavLink: jest.fn().mockImplementation(() => {
      return <></>;
    }),
  };
});

describe("Testing Login component with AuthProvider", () => {
  const loginMock = jest.fn();

  const authContextValue = {
    login: loginMock,
    isLoggingIn: false,
    isRegistering: false,
    user: null,
    register: jest.fn(),
    logout: jest.fn(),
    handlePageLoad: jest.fn(),
  };

  const getUserPreferencesMock = jest.fn();

  const userPreferencesContextValue = {
    getUserPreferences: getUserPreferencesMock,
    createUserPreferences: jest.fn(),
    userPreferences: {
      accepted_tc: new Date(),
      user_id: "0001",
      language: Language.en,
      sessions: [],
    },
    isLoading: false,
  };

  beforeEach(() => {
    // Clear console mocks and mock functions
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
    jest.clearAllMocks();
  });

  beforeAll(() => mockUseTokens());

  test("it should show login form successfully", async () => {
    // Render the component within the AuthContext and Router
    render(
      <HashRouter>
        <AuthContext.Provider value={authContextValue}>
          <UserPreferencesContext.Provider value={userPreferencesContextValue}>
            <Login />
          </UserPreferencesContext.Provider>
        </AuthContext.Provider>
      </HashRouter>
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();

    // AND the component should be rendered
    expect(screen.getByTestId(DATA_TEST_ID.LOGIN_CONTAINER)).toBeDefined();

    // AND the form inputs and button should be displayed
    expect(screen.getByTestId(DATA_TEST_ID.EMAIL_INPUT)).toBeInTheDocument();
    expect(screen.getByTestId(DATA_TEST_ID.PASSWORD_INPUT)).toBeInTheDocument();
    expect(screen.getByTestId(DATA_TEST_ID.LOGIN_BUTTON)).toBeInTheDocument();

    // AND the login button should not be disabled
    expect(screen.queryByTestId(DATA_TEST_ID.LOGIN_BUTTON_CIRCULAR_PROGRESS)).not.toBeInTheDocument();

    // Simulate form input and submission
    fireEvent.change(screen.getByTestId(DATA_TEST_ID.EMAIL_INPUT), { target: { value: "john.doe@example.com" } });
    fireEvent.change(screen.getByTestId(DATA_TEST_ID.PASSWORD_INPUT), { target: { value: "password" } });

    // Trigger form submission
    fireEvent.submit(screen.getByTestId(DATA_TEST_ID.FORM));

    // Expect the login function to have been called
    await waitFor(() => {
      expect(loginMock).toHaveBeenCalledWith(
        "john.doe@example.com",
        "password",
        expect.any(Function),
        expect.any(Function)
      );
    });

    // AND the component should match the snapshot
    expect(screen.getByTestId(DATA_TEST_ID.LOGIN_CONTAINER)).toMatchSnapshot();
  });

  test.each([
    ["accepted", new Date(), routerPaths.ROOT],
    ["not accepted", null, routerPaths.DPA],
  ])(
    "it should handle successful login for a user who has %s terms and conditions",
    async (_description: string, tc: Date | null, expectedPath: string) => {
      // GIVEN the login function will succeed
      const givenUser: TabiyaUser = {
        id: "0001",
        email: "foo@bar.baz",
        name: "Foo Bar",
      };
      loginMock.mockImplementation((email, password, onSuccess, onError) => {
        onSuccess({ id: "0001" });
      });

      // AND the user preferences provider will return the user preferences
      getUserPreferencesMock.mockImplementation((userId, successCallback, failureCallback) => {
        successCallback({ accepted_tc: tc });
      });

      render(
        <HashRouter>
          <AuthContext.Provider value={authContextValue}>
            <UserPreferencesContext.Provider value={userPreferencesContextValue}>
              <Login />
            </UserPreferencesContext.Provider>
          </AuthContext.Provider>
        </HashRouter>
      );

      // THEN expect no errors or warning to have occurred
      expect(console.error).not.toHaveBeenCalled();
      expect(console.warn).not.toHaveBeenCalled();

      // AND the component should be rendered
      expect(screen.getByTestId(DATA_TEST_ID.LOGIN_CONTAINER)).toBeDefined();

      // AND the form inputs and button should be displayed
      expect(screen.getByTestId(DATA_TEST_ID.EMAIL_INPUT)).toBeInTheDocument();
      expect(screen.getByTestId(DATA_TEST_ID.PASSWORD_INPUT)).toBeInTheDocument();
      expect(screen.getByTestId(DATA_TEST_ID.LOGIN_BUTTON)).toBeInTheDocument();

      // Simulate form input and submission
      fireEvent.change(screen.getByTestId(DATA_TEST_ID.EMAIL_INPUT), { target: { value: "foo@bar.baz" } });
      fireEvent.change(screen.getByTestId(DATA_TEST_ID.PASSWORD_INPUT), { target: { value: "password" } });

      // Trigger form submission
      fireEvent.submit(screen.getByTestId(DATA_TEST_ID.FORM));

      // THEN expect the login function to have been called
      await waitFor(() => {
        expect(loginMock).toHaveBeenCalledWith("foo@bar.baz", "password", expect.any(Function), expect.any(Function));
      });

      // AND the user preferences service should have been called
      await waitFor(() => {
        expect(getUserPreferencesMock).toHaveBeenCalledWith(givenUser.id, expect.any(Function), expect.any(Function));
      });
      // AND the user should be redirected to the expected path
      await waitFor(() => {
        expect(useNavigate()).toHaveBeenCalledWith(expectedPath, { replace: true });
      });
    }
  );

  test("it should show an error message if the user's email is not verified", async () => {
    // GIVEN the login function will fail with an error message
    loginMock.mockImplementation((email, password, onSuccess, onError) => {
      const mockServiceError = new ServiceError(
        "AuthService",
        "handleLogin",
        "POST",
        "/signInWithEmailAndPassword",
        StatusCodes.FORBIDDEN,
        ErrorConstants.FirebaseErrorCodes.EMAIL_NOT_VERIFIED,
        "auth/email-not-verified"
      );

      onError(mockServiceError);
    });

    // AND the Login component is rendered within the AuthContext and Router
    render(
      <HashRouter>
        <AuthContext.Provider value={authContextValue}>
          <UserPreferencesContext.Provider value={userPreferencesContextValue}>
            <Login />
          </UserPreferencesContext.Provider>
        </AuthContext.Provider>
      </HashRouter>
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();

    // AND the component should be rendered
    expect(screen.getByTestId(DATA_TEST_ID.LOGIN_CONTAINER)).toBeDefined();

    // AND the form inputs and button should be displayed
    expect(screen.getByTestId(DATA_TEST_ID.EMAIL_INPUT)).toBeInTheDocument();
    expect(screen.getByTestId(DATA_TEST_ID.PASSWORD_INPUT)).toBeInTheDocument();
    expect(screen.getByTestId(DATA_TEST_ID.LOGIN_BUTTON)).toBeInTheDocument();

    // Simulate form input and submission
    fireEvent.change(screen.getByTestId(DATA_TEST_ID.EMAIL_INPUT), { target: { value: "foo@bar.baz" } });
    fireEvent.change(screen.getByTestId(DATA_TEST_ID.PASSWORD_INPUT), { target: { value: "password" } });

    // Trigger form submission
    fireEvent.submit(screen.getByTestId(DATA_TEST_ID.FORM));

    // THEN expect the login function to have been called
    await waitFor(() => {
      expect(loginMock).toHaveBeenCalledWith("foo@bar.baz", "password", expect.any(Function), expect.any(Function));
    });

    // AND the error message should be displayed
    await waitFor(() => {
      expect(useSnackbar().enqueueSnackbar).toHaveBeenCalledWith(
        "The email you are using is registered, but you have not yet verified it. Please verify your email to continue.",
        { variant: "error" }
      );
    });

    // AND the user should not be redirected
    expect(useNavigate()).not.toHaveBeenCalled();
  });

  test("it should show an error message if the user preferences cannot be fetched", async () => {
    // GIVEN the login function will succeed
    const givenUser: TabiyaUser = {
      id: "0001",
      email: "foo@bar.baz",
      name: "Foo Bar",
    };
    loginMock.mockImplementation((email, password, onSuccess, onError) => {
      onSuccess(givenUser);
    });

    // AND the user preferences service will fail
    const givenUserPreferencesError = new ServiceError(
      "ServiceName",
      "ServiceFunction",
      "GET",
      "/api/path",
      StatusCodes.NOT_FOUND,
      ErrorCodes.API_ERROR,
      "Failed to fetch user preferences"
    );
    getUserPreferencesMock.mockImplementation((userId, onSuccess, onError) => {
      onError(givenUserPreferencesError);
    });

    // AND the Login component is rendered within the AuthContext and Router
    jest.spyOn(Logger, "writeServiceErrorToLog");
    render(
      <HashRouter>
        <AuthContext.Provider value={authContextValue}>
          <UserPreferencesContext.Provider value={userPreferencesContextValue}>
            <Login />
          </UserPreferencesContext.Provider>
        </AuthContext.Provider>
      </HashRouter>
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();

    // AND the component should be rendered
    expect(screen.getByTestId(DATA_TEST_ID.LOGIN_CONTAINER)).toBeDefined();

    // AND the form inputs and button should be displayed
    expect(screen.getByTestId(DATA_TEST_ID.EMAIL_INPUT)).toBeInTheDocument();
    expect(screen.getByTestId(DATA_TEST_ID.PASSWORD_INPUT)).toBeInTheDocument();
    expect(screen.getByTestId(DATA_TEST_ID.LOGIN_BUTTON)).toBeInTheDocument();

    // Simulate form input and submission
    fireEvent.change(screen.getByTestId(DATA_TEST_ID.EMAIL_INPUT), { target: { value: "foo@bar.baz" } });
    fireEvent.change(screen.getByTestId(DATA_TEST_ID.PASSWORD_INPUT), { target: { value: "password" } });

    // Trigger form submission
    fireEvent.submit(screen.getByTestId(DATA_TEST_ID.FORM));

    // THEN expect the login function to have been called
    await waitFor(() => {
      expect(loginMock).toHaveBeenCalledWith("foo@bar.baz", "password", expect.any(Function), expect.any(Function));
    });

    // AND the user preferences service should have been called
    await waitFor(() => {
      expect(getUserPreferencesMock).toHaveBeenCalledWith(givenUser.id, expect.any(Function), expect.any(Function));
    });

    // AND the error message should be displayed
    await waitFor(() => {
      expect(useSnackbar().enqueueSnackbar).toHaveBeenCalledWith(
        "The requested resource was not found. Please clear your browser's cache and refresh the page.",
        {
          variant: "error",
        }
      );
    });

    // AND the error should be logged
    expect(Logger.writeServiceErrorToLog).toHaveBeenCalledWith(givenUserPreferencesError, console.error);

    // AND the user should not be redirected
    expect(useNavigate()).not.toHaveBeenCalled();
  });

  test("it should show error message when login fails", async () => {
    // GIVEN the login function will fail
    loginMock.mockImplementation((email, password, onSuccess, onError) => {
      onError(new Error("Login failed"));
    });

    // AND the Login component is rendered within the AuthContext and Router
    render(
      <HashRouter>
        <AuthContext.Provider value={authContextValue}>
          <Login />
        </AuthContext.Provider>
      </HashRouter>
    );

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();

    // AND the component should be rendered
    expect(screen.getByTestId(DATA_TEST_ID.LOGIN_CONTAINER)).toBeDefined();

    // AND the form inputs and button should be displayed
    expect(screen.getByTestId(DATA_TEST_ID.EMAIL_INPUT)).toBeInTheDocument();
    expect(screen.getByTestId(DATA_TEST_ID.PASSWORD_INPUT)).toBeInTheDocument();
    expect(screen.getByTestId(DATA_TEST_ID.LOGIN_BUTTON)).toBeInTheDocument();

    // Simulate form input and submission
    fireEvent.change(screen.getByTestId(DATA_TEST_ID.EMAIL_INPUT), { target: { value: "foo@bar.baz" } });
    fireEvent.change(screen.getByTestId(DATA_TEST_ID.PASSWORD_INPUT), { target: { value: "password" } });

    // Trigger form submission
    fireEvent.submit(screen.getByTestId(DATA_TEST_ID.FORM));

    // THEN expect the login function to have been called
    await waitFor(() => {
      expect(loginMock).toHaveBeenCalledWith("foo@bar.baz", "password", expect.any(Function), expect.any(Function));
    });

    // AND the error message should be displayed
    await waitFor(() => {
      expect(useSnackbar().enqueueSnackbar).toHaveBeenCalledWith(
        "An unexpected error occurred. Please try again later.",
        { variant: "error" }
      );
    });
  });

  it("should disable everything when logging in", () => {
    // GIVEN the isLoggingIn flag is set to true
    const _authContextValue = {
      ...authContextValue,
      isLoggingIn: true,
    };

    // WHEN the Login component is rendered within the AuthContext and Router
    render(
      <HashRouter>
        <AuthContext.Provider value={_authContextValue}>
          <Login />
        </AuthContext.Provider>
      </HashRouter>
    );

    // THEN expect the login button to be disabled
    expect(screen.getByTestId(DATA_TEST_ID.LOGIN_BUTTON)).toBeDisabled();
    // AND the email input to be disabled
    expect(screen.getByTestId(DATA_TEST_ID.EMAIL_INPUT)).toBeDisabled();
    // AND the password input to be disabled
    expect(screen.getByTestId(DATA_TEST_ID.PASSWORD_INPUT)).toBeDisabled();

    // AND login button circular progress to be displayed
    expect(screen.getByTestId(DATA_TEST_ID.LOGIN_BUTTON_CIRCULAR_PROGRESS)).toBeInTheDocument();
  });
});
