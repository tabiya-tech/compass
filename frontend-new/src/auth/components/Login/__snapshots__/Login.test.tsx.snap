// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Login component with AuthProvider it should show login form successfully 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthXs css-xravbe-MuiContainer-root"
  data-testid="login-container-7ce9ba1f-bde0-48e2-88df-e4f697945cc4"
>
  <div
    class="MuiBox-root css-inbm3t"
  >
    <div
      class="MuiBox-root css-1vd84sn"
    >
      <img
        alt="Logo"
        data-testid="login-logo-7ce9ba1f-bde0-48e2-88df-e4f697945cc4"
        src="/logo.svg"
        style="max-width: 60%; margin: 10%;"
      />
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary MuiIconButton-sizeMedium css-j7z1rr-MuiButtonBase-root-MuiIconButton-root"
        data-testid="auth-language-selector-f4d06e4b-0e0c-49c7-ad93-924c5ac89070"
        tabindex="0"
        title="Language Selector"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-etbowg-MuiSvgIcon-root"
          data-testid="LanguageOutlinedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M11.99 2C6.47 2 2 6.48 2 12s4.47 10 9.99 10C17.52 22 22 17.52 22 12S17.52 2 11.99 2m6.93 6h-2.95c-.32-1.25-.78-2.45-1.38-3.56 1.84.63 3.37 1.91 4.33 3.56M12 4.04c.83 1.2 1.48 2.53 1.91 3.96h-3.82c.43-1.43 1.08-2.76 1.91-3.96M4.26 14C4.1 13.36 4 12.69 4 12s.1-1.36.26-2h3.38c-.08.66-.14 1.32-.14 2s.06 1.34.14 2zm.82 2h2.95c.32 1.25.78 2.45 1.38 3.56-1.84-.63-3.37-1.9-4.33-3.56m2.95-8H5.08c.96-1.66 2.49-2.93 4.33-3.56C8.81 5.55 8.35 6.75 8.03 8M12 19.96c-.83-1.2-1.48-2.53-1.91-3.96h3.82c-.43 1.43-1.08 2.76-1.91 3.96M14.34 14H9.66c-.09-.66-.16-1.32-.16-2s.07-1.35.16-2h4.68c.09.65.16 1.32.16 2s-.07 1.34-.16 2m.25 5.56c.6-1.11 1.06-2.31 1.38-3.56h2.95c-.96 1.65-2.49 2.93-4.33 3.56M16.36 14c.08-.66.14-1.32.14-2s-.06-1.34-.14-2h3.38c.16.64.26 1.31.26 2s-.1 1.36-.26 2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    <h4
      class="MuiTypography-root MuiTypography-h4 MuiTypography-gutterBottom css-xvfkrl-MuiTypography-root"
      data-testid="login-title-7ce9ba1f-bde0-48e2-88df-e4f697945cc4"
    >
      Welcome to Compass!
    </h4>
    <p
      class="MuiTypography-root MuiTypography-body2 MuiTypography-gutterBottom css-1ncy5w-MuiTypography-root"
      data-testid="login-subtitle-7ce9ba1f-bde0-48e2-88df-e4f697945cc4"
    >
      Login to your account to continue
    </p>
    <form
      class="MuiBox-root css-1c88hr3"
      data-testid="login-form-7ce9ba1f-bde0-48e2-88df-e4f697945cc4"
    >
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-11i71sc-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r0:"
          id=":r0:-label"
        >
          Email
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1th3aq5-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-no50bp-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1apxpsd-MuiInputBase-input-MuiOutlinedInput-input"
            data-testid="login-email-input-7ce9ba1f-bde0-48e2-88df-e4f697945cc4"
            id=":r0:"
            required=""
            type="email"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Email
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-11i71sc-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          for=":r1:"
          id=":r1:-label"
        >
          Password
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-1th3aq5-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-no50bp-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1apxpsd-MuiInputBase-input-MuiOutlinedInput-input"
            data-testid="login-password-input-7ce9ba1f-bde0-48e2-88df-e4f697945cc4"
            id=":r1:"
            required=""
            type="password"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Password
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation MuiButton-fullWidth MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-disableElevation MuiButton-fullWidth css-q25fil-MuiButtonBase-root-MuiButton-root"
        data-testid="login-button-7ce9ba1f-bde0-48e2-88df-e4f697945cc4"
        style="margin-top: 16px;"
        tabindex="0"
        type="submit"
      >
        Login
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </form>
    <div
      class="MuiBox-root css-9kk22j"
      data-testid="firebase-auth-container-7ce9ba1f-bde0-48e2-88df-e4f697945cc4"
    >
      <p
        class="MuiTypography-root MuiTypography-body2 css-1t5l3lb-MuiTypography-root"
        data-testid="login-using-7ce9ba1f-bde0-48e2-88df-e4f697945cc4"
      >
        Or continue with
      </p>
      <div
        class="MuiBox-root css-lyhpgt"
      >
        <span />
      </div>
    </div>
    <p
      class="MuiTypography-root MuiTypography-body2 css-1t5l3lb-MuiTypography-root"
      data-testid="login-login-link-7ce9ba1f-bde0-48e2-88df-e4f697945cc4"
    >
      Don't have an account?
       
    </p>
  </div>
</div>
`;
